### ISSUE TITLE
New leader takes too long to reprocess

### ISSUE THREAD

**Issue by deepthidevaki** [zeebe-io/zeebe#3107]
**Describe the bug**

New leader takes too long to be ready, when a fail over occurs after the cluster is running for a long time. 

**To Reproduce**
Can probably reproduce it by leaving the cluster running for a long time with high load.

**Expected behavior**
Fail over should be fast.

**Log/Stacktrace**
In the following case it took ~20 min to reprocess.

<details><summary>Full Stacktrace</summary>
 <p>

```
I 2019-09-24T15:05:31.305435134Z 15:05:31.305 [partition-5-processor] [zeebe-0.zeebe.deepthi.svc.cluster.local:26501-zb-actors-1] DEBUG io.zeebe.logstreams - Recovering state of partition 5 from snapshot 
I 2019-09-24T15:08:07.937078780Z 15:08:07.935 [partition-5-processor] [zeebe-0.zeebe.deepthi.svc.cluster.local:26501-zb-actors-1] INFO  io.zeebe.logstreams - Recovered state of partition 5 from snapshot at position 1043678960576 
I 2019-09-24T15:08:07.940212069Z 15:08:07.939 [partition-5-processor-snapshot-director] [zeebe-0.zeebe.deepthi.svc.cluster.local:26501-zb-actors-0] DEBUG io.zeebe.logstreams.snapshot - The position of the last valid snapshot is '1039387668480'. Taking snapshots beyond this position. 
I 2019-09-24T15:18:17.099224299Z 15:18:17.098 [partition-5-processor] [zeebe-0.zeebe.deepthi.svc.cluster.local:26501-zb-actors-1] INFO  io.zeebe.processor - Processor starts reprocessing, until last source event position 2229095282968 
I 2019-09-24T15:23:08.055131952Z 15:23:08.054 [partition-5-processor-snapshot-director] [zeebe-0.zeebe.deepthi.svc.cluster.local:26501-zb-fs-workers-0] DEBUG io.zeebe.logstreams.snapshot - No changes since last snapshot we will skip snapshot creation. Last valid snapshot position 1039387668480, new lower bound position -1 
I 2019-09-24T15:26:08.818584465Z 15:26:08.817 [partition-5-processor] [zeebe-0.zeebe.deepthi.svc.cluster.local:26501-zb-actors-0] INFO  io.zeebe.processor - Processor finished reprocessing at event position 2229095282968
```

</p>
</details>

**Environment:**
 - OS: linux, k8s
 - Zeebe Version: 0.21.0-alpha3

**Comment by Zelldon** [comment]
Hey @deepthidevaki I assume this is a duplicate of https://github.com/zeebe-io/zeebe/issues/3028

**Comment by deepthidevaki** [comment]
@Zelldon Not sure. Isn't that one more related to rejoin the cluster after restart? In this issue, I observed the problems in the new leader which was already in the cluster and successfully replicating events. Atleast, in the first look, it looks like an issue in the stream processor rather than in Atomix.

**Comment by Zelldon** [comment]
I think you're right sorry.

**Comment by deepthidevaki** [comment]
**Running local, 1 partition, 1 broker**
Restarted the broker when it has 24 segments, 1 snapshot, around 5 min processing after snapshot. It was fast to seek even though there are many segments.
Seek time: 10s
Scan log: 15s
Seek time: 10s
Reprocessing: 1m10s
<details><summary>Logs</summary>
 <p>

```
13:07:37.856 Recovering state of partition 1 from snapshot
13:07:37.896 Recovering state of partition 1 : seek to snapshot position 4874793564208
13:07:47.817 Recovering state of partition 1 : seeked to snapshot position 4874793564208
13:07:47.818 Recovered state of partition 1 from snapshot at position 4874793564208
13:07:48.019 Start scanning the log for error events.
13:08:03.114 Finished scanning the log for error events.
13:08:03.114 Processor seeking to snapshot position 4874793564208
13:08:15.539 Processor starts reprocessing, until last source event position 6524058836672
13:09:23.953 Processor finished reprocessing at event position 6524058836672
```

</p>
</details>

**Running on cloud**
**Case 1**: SnapshotPeriod 15min. Last snapshot created around 11.43 (14 min ago)
Seek time: ~1m
Scan log: 1m
Seek time: 1s
Reprocessing: 5m
<details><summary>Logs</summary>
 <p>

```
11:57:29.800 [partition-1-processor] - Recovering state of partition 1 from snapshot 
11:57:29.801 [partition-1-processor] - Seek to snapshot position 1017908113528 
11:58:22.199 [partition-1-processor] - Finished seeking to snapshot position 1017908113528 
11:58:22.200 [partition-1-processor] - Recovered state of partition 1 from snapshot at position 1017908113528
 
11:58:22.449 [partition-1-processor] - Start scanning the log for error events. 
11:59:26.931 [partition-1-processor] - Finished scanning the log for error events.

11:59:26.932 [partition-1-processor] - Processor seeking to snapshot position 1017908113528 
11:59:27.556 [partition-1-processor] - Finished seeking to snapshot position 1017908113528
 
11:59:27.556 [partition-1-processor] - Processor starts reprocessing, until last source event position 1911264584424 
12:04:53.512 [partition-1-processor] - Processor finished reprocessing at event position 1911264584424
```

</p>
</details>

**Case 2**: SnapshotPeriod 15min. 3 snapshots (from around 45 minutes of processing) Compacted only 2 segments, have segments 3 - 18

Seek time: 8m
Scan log: 5m
Seek time: 6m
Reprocessing: 1m25s
<details><summary>Logs</summary>
 <p>

```
12:28:21.217 [partition-3-processor] - Recovering state of partition 3 from snapshot
 
12:28:21.218 [partition-3-processor] - Seek to snapshot position 4389464546736 
12:36:52.277 [partition-3-processor] - Finished seeking to snapshot position 4389464546736
 
12:36:52.298 [partition-3-processor] - Recovered state of partition 3 from snapshot at position 4389464546736
 
12:36:52.300 [partition-3-processor] - Start scanning the log for error events.
12:41:48.304 [partition-3-processor] - Finished scanning the log for error events.
 
12:41:48.304 [partition-3-processor] - Processor seeking to snapshot position 4389464546736
12:48:18.426 [partition-3-processor] - Finished seeking to snapshot position 4389464546736
 
12:48:18.426 [partition-3-processor] - Processor starts reprocessing, until last source event position 5171148344072
12:49:45.494 [partition-3-processor] - Processor finished reprocessing at event position 5171148344072
```

</p>
</details>

**Case 3**: SnapshotPeriod 5m.

It looks much better.
Seek time: 48s
Scan: 2s
Reprocessing time ~1m,
<details><summary>Logs</summary>
 <p>

```
14:12:00.322 [partition-3-processor] - Recovering state of partition 3 from snapshot
 
14:12:00.322 [partition-3-processor] - Seek to snapshot position 2126013179616 
14:12:48.150 [partition-3-processor] - Finished seeking to snapshot position 2126013179616
 
14:12:48.150 [partition-3-processor] - Recovered state of partition 3 from snapshot at position 2126013179616
 
14:12:48.152 [partition-3-processor] - Start scanning the log for error events.
14:12:50.172 [partition-3-processor] - Finished scanning the log for error events.
 
14:12:50.173 [partition-3-processor] - Processor seeking to snapshot position 2126013179616
14:12:50.660 [partition-3-processor] - Finished seeking to snapshot position 2126013179616
 
14:12:50.660 [partition-3-processor] - Processor starts reprocessing, until last source event position 2452428959352
14:13:44.828 [partition-3-processor] - Processor finished reprocessing at event position 2452428959352
```

</p>
</details>

**Observations**
* Reprocessing takes around 1/3 of processing time. 
* Seek takes too much time if the logstream is large.
* It helps to reduce the snapshotPeriod, so that we compact more often and there is less to reprocess.
* Seek must be optimized.

**Comment by deepthidevaki** [comment]
Fixed by #3131
